src/lib/protect/apply-patch.js

```diff
@@ -29,48 +26,84 @@ function applyPatch(patch, vuln, live) {
       debug('Failed loading package.json of package about to be patched', err);
     }
 
-    // default is 200K, we increase it as we've seen
-    //  "stdout maxBuffer exceeded" erros in some user debug logs
-    var STDOUT_BUFFER_SIZE = 5 * 1024 * 1024;
+    var patchContent = fs.readFileSync(path.resolve(relative, patch), 'utf8');
 
-    // do a dry run first, otherwise the patch can "succeed" (exit 0) if it
-    // only manages to patch *some* of the chunks (and leave the file partly
-    // patched).
-    exec(cmd + test, {
-      cwd: cwd,
-      env: process.env,
-      maxBuffer: STDOUT_BUFFER_SIZE,
-    }, function (error, stdout) { // stderr is ignored
-      var out = stdout.trim();
-      if (error || out.indexOf('FAILED') !== -1) {
-        debug('patch command failed', relative, error, out);
-        return patchError(error, out, relative, vuln).catch(reject);
-      }
-      if (!live) {
-        return resolve();
-      }
+    jsDiff(patchContent, relative, live).then(function () {
+      debug('patch succeed');
+      resolve();
+    }).catch(function (error) {
+      debug('patch command failed', relative, error);
+      patchError(error, relative, vuln).catch(reject);
+    });
+  });
+}
 
-      // if it was okay, and it wasn't a dry-run, then let's do it for real
-      exec(cmd, {
-        cwd: cwd,
-        env: process.env,
-        maxBuffer: STDOUT_BUFFER_SIZE,
-      }, function (error, stdout) {
-        var out = stdout.trim();
-        if (error || out.indexOf('FAILED') !== -1) {
-          debug('patch command failed', relative, error, out);
-          return patchError(error, out, relative, vuln).catch(reject);
+function jsDiff(patchContent, relative, live) {
+  const files = {};
+  return new Promise(function (resolve, reject) {
+    diff.applyPatches(patchContent, {
+      loadFile: function (index, callback) {
+        try {
+          var fileName = stripFirstSlash(index.oldFileName);
+          if (files[fileName]) {
+            return callback(null, files[fileName]);
+          }
+          var content = fs.readFileSync(path.resolve(relative, fileName), 'utf8');
+          callback(null, content);
+        } catch (err) {
+          callback(err);
         }
-
-        debug('patch succeed', out);
-
-        resolve();
-      });
+      },
+      patched: function (index, content, callback) {
+        try {
+          if (content === false) {
+            throw new Error('Found a mismatching patch\n' + JSON.stringify(index, null, 2));
+          }
+          var newFileName = stripFirstSlash(index.newFileName);
+          var oldFileName = stripFirstSlash(index.oldFileName);
+          if (newFileName !== oldFileName) {
+            files[oldFileName] = null;
+          }
+          files[newFileName] = content;
+          callback();
+        } catch (err) {
+          callback(err);
+        }
+      },
+      compareLine: function (_, line, operation, patchContent) {
+        if (operation === ' ') {
```

Comment:
a comment will help here. what is operation `' '` ?

Reply:
`' '` means there is no operation. I needed this because jsdiff by default compare all lines and fails when an operation-less diff line mismatches, which differs from GNU patch behavior which only checks for `+`/`-`/`!`.

Reply:
The comment would be:

```suggestion
        if (operation === ' ') {
          // Ignore when no patch operators as GNU patch does
```